src/main/kotlin/org/rust/ide/presentation/Utils.kt

```diff
@@ -116,17 +114,80 @@ fun breadcrumbName(e: RsElement): String? {
         is RsImplItem -> {
             val typeName = run {
                 val typeReference = e.typeReference
-                (typeReference?.skipParens() as? RsBaseType)?.path?.let { lastComponentWithoutGenerics(it) }
+                (typeReference?.skipParens() as? RsBaseType)?.path?.referenceName
                     ?: typeReference?.text
                     ?: return null
             }
 
-            val traitName = e.traitRef?.path?.let { lastComponentWithoutGenerics(it) }
+            val traitName = e.traitRef?.path?.referenceName
             val start = if (traitName != null) "$traitName for" else "impl"
             "$start $typeName"
         }
 
         is RsFunction -> e.name?.let { "$it()" }
+
+        is RsIfExpr -> {
+            buildString {
+                append("if")
+
+                val condition = e.condition
+                if (condition != null) {
+                    if (condition.expr is RsBlockExpr)
+                        append(" {...}")
+                    else
+                        append(' ').append(condition.text)
+                }
+            }
+        }
+        is RsElseBranch -> e.`else`.text
+
+        is RsLoopExpr -> e.loop.text
+
+        is RsForExpr -> {
+            buildString {
+                if (e.labelDecl != null) {
+                    append(e.labelDecl?.text).append(' ')
+                }
+
+                append("for")
+
+                if (e.block != null) {
+                    append(' ').append(e.pat?.text)
+                    append(" in ").append(e.expr?.text)
+                }
+            }
+        }
+
+        is RsWhileExpr -> {
+            buildString {
+                append("while")
+
+                val condition = e.condition
+                if (condition != null) {
+                    if (condition.expr is RsBlockExpr)
```

Comment:
Do we really need this condition?
For example, it doesn't work in the following case:
```rust
if ({ 1 > 2 }) {}
```

because top-level expression in condition is not block expr

Reply:
This judgment is mainly used in the case of partial expressions generated after keyword completion.

```rust
fn main() {
    if /*caret*/ { }
}
```